src/arrays.jl

```diff
@@ -0,0 +1,177 @@
+
+function Base.BroadcastStyle(::Type{QA}) where {QA<:QuantityArray}
+    return Broadcast.ArrayStyle{QA}()
+end
+function Base.BroadcastStyle(::Broadcast.ArrayStyle{QA1}, ::Broadcast.ArrayStyle{QA2}) where {QA1<:QuantityArray,QA2<:QuantityArray}
+    return Broadcast.ArrayStyle{promote_type(QA1, QA2)}()
+end
+function Base.similar(bc::Broadcast.Broadcasted{Broadcast.ArrayStyle{QA}}, ::Type{ElType}) where {QA<:QuantityArray,ElType}
+    output_array_type = constructor_of(array_type(QA)){unwrap_quantity(ElType)}
+    output_array = similar(output_array_type, axes(bc))
+
+    if ElType <: AbstractQuantity
+        first_output = materialize_first(bc)
+        if typeof(first_output) != ElType
+            @warn (
+                "Materialization of first element likely failed. "
+                * "Please submit a bug report with information on "
+                * "the function you are broadcasting."
+            )
+        end
+        return QuantityArray(output_array, dimension(first_output), ElType)
+    else
+        return output_array
+    end
+end
+unwrap_quantity(::Type{Q}) where {T,Q<:AbstractQuantity{T}} = T
+unwrap_quantity(::Type{T}) where {T} = T
+
+# Basically, we want to solve a single element to find the output dimension.
+# Then we can put results in the output `QuantityArray`.
+materialize_first(bc::Base.Broadcast.Broadcasted) = bc.f(materialize_first.(bc.args)...)
+
+# Base cases
+materialize_first(q::AbstractQuantity) = q
+materialize_first(q::AbstractQuantity, ::Any) = q
+materialize_first(q::QuantityArray) = first(q)
+materialize_first(q::QuantityArray, ::Any) = first(q)
+materialize_first(q::AbstractArray{Q}) where {Q<:AbstractQuantity} = first(q)
+materialize_first(q::AbstractArray{Q}, ::Any) where {Q<:AbstractQuantity} = first(q)
+
+# Derived calls
+materialize_first(r::Base.RefValue) = materialize_first(r.x)
+materialize_first(x::Base.Broadcast.Extruded) = materialize_first(x.x)
+materialize_first(args::Tuple) = materialize_first(first(args), Base.tail(args))
+materialize_first(args::AbstractArray) = length(args) >= 1 ? materialize_first(args[begin], args[begin+1:end]) : error("Unexpected broadcast format. Please submit a bug report.")
+materialize_first(::Tuple{}) = error("Unexpected broadcast format. Please submit a bug report.")
+materialize_first(::Any, rest) = materialize_first(rest)
+
+# Everything else:
+materialize_first(x) = x
+
+function _print_array_type(io::IO, ::Type{QA}) where {QA<:QuantityArray}
+    return print(io, "QuantityArray(::", array_type(QA), ", ::", quantity_type(QA), ")")
+end
+Base.showarg(io::IO, v::QuantityArray, _) = _print_array_type(io, typeof(v))
+Base.show(io::IO, ::MIME"text/plain", ::Type{QA}) where {QA<:QuantityArray} = _print_array_type(io, QA)
+
+# Other array operations:
+Base.copy(A::QuantityArray) = QuantityArray(copy(ustrip(A)), dimension(A), quantity_type(A))
```

Comment:
Should copy dimensions too, as some dimensions (SymbolicDimensions) store their data with an array